core_debug.tex

```diff
@@ -37,8 +37,10 @@ \section{Debug Mode} \label{debugmode}
     and {\tt dret}.  (To change the privilege level, the debugger can write
     \FcsrDcsrPrv and \FcsrDcsrV in \RcsrDcsr). The only exception is {\tt ebreak}, which ends
     execution of the Program Buffer when executed.
-\item \label{fence} Completing Program Buffer execution is considered output for the purpose
-    of {\tt fence} instructions.
+\item \label{fence} Completing Program Buffer execution is considered output
+    for the purpose of {\tt fence} instructions.
+    This includes executing an {\tt ebreak} or {\tt dret} instruction, and
+    any exceptions that may be encounterd.
```

Comment:
The act of clearing abstractcs.busy indicates PB completion, not the execution of these instructions.  I think that the rule should refer to the act of clearing that bit rather than this proposed sentence for a few reasons:
1. The sentence "includes" those things but is not necessarily exhaustive.  Although I can't think of any other cases, clearing abstractcs.busy includes these 3 cases and any others that we may be missing.
2. If an execution-based implementation executes a dret in the PB then it's UNSPECIFIED so arguably this sentence doesn't apply (or maybe it does if the dret actually terminates PB execution).  It just seems cleaner to talk about the observable act of abstractcs.busy becoming 0 than having to know whether rule 8 overrides rule 9.

Reply:
Hi @timsifive ,
I am still a bit confused about the need to have this rule. Any 'device output' or 'memory write' should follow the rules for the FENCE instructions, whether or not the hart is in normal mode, debug mode, or running code from PB or any other location. If there is no exception to following the FENCE instruction as defined in the main RISC-V specification, then re-stating it here seems confusing to me. Perhaps you are suggesting that FENCE is not valid within PB execution but is valid only when finishing PB execution, which I doubt is the intention.

If though you are suggesting that completing the program buffer will automatically invoke a FENCE operation then that rule should be stated differently. But I don't think you are suggesting this behavior either.

Perhaps you are wanting to give a rule (or a note or a suggestion) that the debug (rom) code shall(..or shoud) include a FENCE instruction as the first instruction to be executed.

If you plan to keep the existing rule, could you consider changing the word " output " to " 'device output' " or '" memory write' " to be inline with the FENCE instruction terminology?

Please also refer to the comments posted on the RISC-V tech-debug listserve with the email subject title "clarification request: completion of program buffer is output for fence". One comment was that this rule should be more of a note.